### ISSUE TITLE
Race condition or stale variable in SslStreamInternal causing System.NotSupportedException in SslStream.Write

### ISSUE THREAD

**Issue by pieter-venter** [dotnet/corefx#36076]
Over the past 8 months, we have observed the following Exception from time to time in production:

``` C#
Exception: System.NotSupportedException EX8C9B2342
Message: The WriteAsync method cannot be called when another write operation is pending.
StackTrace: 
at System.Net.Security.SslStreamInternal.WriteAsyncInternal[TWriteAdapter](TWriteAdapter writeAdapter, ReadOnlyMemory1 buffer)
at System.Net.Security.SslStreamInternal.Write(Byte[] buffer, Int32 offset, Int32 count)
at System.Net.Security.SslStream.Write(Byte[] buffer, Int32 offset, Int32 count)
```
We confirmed that all writes to the SslStream are properly synchronized using a lock(). We have a specific object used to synchronize on and then issue a write on the SslStream via BinaryWriter **only** while holding the lock. We will never be issuing multiple writes on the same SslStream.

Not the actual code, but in principle this:

``` C#
// This shared object is always used to synchronize on
readonly object networkStreamLock = new object();

// At some point we connect and initialize SslStream
Stream networkStream = new SslStream(...);

// Create writer to make simple to write primitives to stream
writer = new BinaryWriter(networkStream);

// All writes are done Synchronously (but could be from different threads) and follow this pattern.
// This is where we observe the Exception.
lock (networkStreamLock)
{
    writer.Write( /* some value */);
}
```
**Some context:**
We have close to a thousand servers in production running .Net Core, with each server having between 4 and 64 cores. So it's quite a large deployment and this error is relatively rare but measurable.

We log all errors in production and our analysis has shown the following:

![ssl-errors-by-month](https://user-images.githubusercontent.com/12296644/54447123-fcf4bf80-471e-11e9-8bfa-e8b519e1a650.png)

The first instance of this error happened on 4 July 2018. Our deployment history shows this is also when we rolled out version 2.1 of .Net Core. 
I.e. this issue first started appearing when we went from **2.0 -> 2.1** of .Net Core and continues to happen in today (running the latest .Net Core 2.2.x).

Looking at the [history of SslStreamInternal](https://github.com/dotnet/corefx/commits/master/src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs), @stephentoub and others implemented async reads and writes into version 2.1, with the synchronous versions calling the async equivalent internally like this:

``` C#
internal void Write(byte[] buffer, int offset, int count)
{
    ValidateParameters(buffer, offset, count);

    SslWriteSync writeAdapter = new SslWriteSync(_sslState);
    WriteAsyncInternal(writeAdapter, new ReadOnlyMemory<byte>(buffer, offset, count)).GetAwaiter().GetResult();
}
```

The error we observed is thrown [here](https://github.com/dotnet/corefx/blob/8681eefb557edbb50bd9e96c785388d836d44331/src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs#L353):
``` C#
if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
 {
    throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(WriteAsync), "write"));
}
```
Notice the use of Interlocked.Exchange to compare atomically if the value is already 1.

However, this value assigned a value of 0 once the write completes (or exception happened) [like so](https://github.com/dotnet/corefx/blob/8681eefb557edbb50bd9e96c785388d836d44331/src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs#L362):
``` C#
_nestedWrite = 0;
```
Notice the lack of Volatile.Write, declaring the variable as volatile or memory barrier.

**Hypothesis**:
TLDR; The asymmetry in using Interlock.Exchange when setting the value to 1 but no synchronization when setting to 0 could be a problem.

Now I'm probably out of my depth here, but it seems to me what is happening is that on a multi core server, on very rare occasions, we are seeing a CPU cache coherence issue. Where one CPU still observes the value of 1 while another set it to 0 (without proper fencing ensuring caches are flushed). Since the Write method now internally uses async, it seems reasonable that different CPUs can be involved when executing the write (switching at `awaits` as the scheduler deems fit) and therefore the load and set operations for _nestedWrite can happen on different CPUs. 

Some useful stuff here:
https://afana.me/archive/2015/07/10/memory-barriers-in-dot-net.aspx/

Unfortunately I have been unable to reproduce this with a small reproduction. Given the size of our deployment and the relatively rare occurrence of this - it does seem to be consistent with some nasty multi-threading/jit optimization/cpu instruction reordering issue.

cc
@baal2000 for his help in reviewing this issue.

**Comment by stephentoub** [comment]
> this issue first started appearing when we went from 2.0 -> 2.1 of .Net Core

While it's possible it's related to a change in SslStream, it's also quite possible it's related to performance improvements elsewhere in the stack.  We made a lot of stuff faster going from 2.0 to 2.1, in particular around networking, so it's possible some latent race condition (in the core libraries or in your code) is just showing up as a result of the race condition becoming more prominent.

> The asymmetry in using Interlock.Exchange when setting the value to 1 but no synchronization when setting to 0 could be a problem.

It shouldn't be.  If the consuming code is written correctly, then some other synchronization would be in place (e.g. your locks) to ensure that a secondary read isn't issued until the first has completed, and that synchronization would ensure that there's no problem with this pattern.  If your code is actually correctly locking on the correct lock around every call, then even the interlocked exchange is overkill: no one outside of your lock should ever even touch that field, no writes to that field will move outside of the lock, and only code that holds the lock will examine the field.

**Comment by stephentoub** [comment]
I've been looking over the code.

@pieter-venter, is it possible this happens (and only happens) in your project after a call has done a write that's failed with an exception? i.e. one thread calls Write, it throws, and then either the same code or some other code does another Write on the same instance?

**Comment by pieter-venter** [comment]
Let me look into this and see if we have evidence of exceptions preceding this one.
